File: frontend/src/app/services/tarefa.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Tarefa, StatusTarefa, TarefaCreate, TarefaUpdate } from '../models/tarefa.model';

/**
 * Serviço para gerenciar operações relacionadas a tarefas.
 *
 * Este serviço fornece métodos para comunicação com a API REST
 * do backend para CRUD de tarefas.
 *
 * @author Stefanini Challenge
 * @version 1.0
 * @since 2026-01-06
 */
@Injectable({
  providedIn: 'root'
})
export class TarefaService {
  private apiUrl = 'http://localhost:8080/api/tarefas';

  constructor(private http: HttpClient) {}

  /**
   * Lista todas as tarefas.
   * @returns Observable com array de tarefas
   */
  listarTodas(): Observable<Tarefa[]> {
    return this.http.get<Tarefa[]>(this.apiUrl);
  }

  /**
   * Lista tarefas filtradas por status.
   * @param status Status das tarefas a serem listadas
   * @returns Observable com array de tarefas
   */
  listarPorStatus(status: StatusTarefa): Observable<Tarefa[]> {
    return this.http.get<Tarefa[]>(`${this.apiUrl}/status/${status}`);
  }

  /**
   * Busca uma tarefa por ID.
   * @param id ID da tarefa
   * @returns Observable com a tarefa encontrada
   */
  buscarPorId(id: number): Observable<Tarefa> {
    return this.http.get<Tarefa>(`${this.apiUrl}/${id}`);
  }

  /**
   * Cria uma nova tarefa.
   * @param tarefa Dados da tarefa a ser criada
   * @returns Observable com a tarefa criada
   */
  criar(tarefa: TarefaCreate): Observable<Tarefa> {
    return this.http.post<Tarefa>(this.apiUrl, tarefa);
  }

  /**
   * Atualiza uma tarefa existente.
   * @param id ID da tarefa a ser atualizada
   * @param tarefa Dados atualizados da tarefa
   * @returns Observable com a tarefa atualizada
   */
  atualizar(id: number, tarefa: TarefaUpdate): Observable<Tarefa> {
    return this.http.put<Tarefa>(`${this.apiUrl}/${id}`, tarefa);
  }

  /**
   * Deleta uma tarefa.
   * @param id ID da tarefa a ser deletada
   * @returns Observable vazio
   */
  deletar(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}

File: backend/src/main/java/com/stefanini/todolist/application/services/TarefaService.java
package com.stefanini.todolist.application.services;

import com.stefanini.todolist.application.dtos.TarefaCreateDTO;
import com.stefanini.todolist.application.dtos.TarefaResponseDTO;
import com.stefanini.todolist.application.dtos.TarefaUpdateDTO;
import com.stefanini.todolist.domain.enums.StatusTarefa;

import java.util.List;

/**
 * Interface de serviço para operações de negócio relacionadas a Tarefas.
 *
 * <p>Define o contrato para as operações CRUD e funcionalidades adicionais
 * do sistema de gerenciamento de tarefas.</p>
 *
 * <p>Esta interface segue o princípio de Inversão de Dependência (SOLID),
 * permitindo que implementações sejam facilmente substituídas.</p>
 *
 * @author Stefanini Challenge
 * @version 1.0
 * @since 2026-01-06
 */
public interface TarefaService {

    /**
     * Cria uma nova tarefa no sistema.
     *
     * @param createDTO Dados da tarefa a ser criada
     * @return DTO com os dados da tarefa criada
     */
    TarefaResponseDTO criar(TarefaCreateDTO createDTO);

    /**
     * Busca uma tarefa por seu ID.
     *
     * @param id Identificador único da tarefa
     * @return DTO com os dados da tarefa encontrada
     * @throws com.stefanini.todolist.infrastructure.exceptions.ResourceNotFoundException
     *         se a tarefa não for encontrada
     */
    TarefaResponseDTO buscarPorId(Long id);

    /**
     * Lista todas as tarefas cadastradas.
     *
     * @return Lista com todas as tarefas
     */
    List<TarefaResponseDTO> listarTodas();

    /**
     * Lista tarefas filtradas por status.
     *
     * @param status Status das tarefas a serem listadas
     * @return Lista de tarefas com o status especificado
     */
    List<TarefaResponseDTO> listarPorStatus(StatusTarefa status);

    /**
     * Atualiza os dados de uma tarefa existente.
     *
     * @param id Identificador único da tarefa
     * @param updateDTO Novos dados da tarefa
     * @return DTO com os dados da tarefa atualizada
     * @throws com.stefanini.todolist.infrastructure.exceptions.ResourceNotFoundException
     *         se a tarefa não for encontrada
     */
    TarefaResponseDTO atualizar(Long id, TarefaUpdateDTO updateDTO);

    /**
     * Remove uma tarefa do sistema.
     *
     * @param id Identificador único da tarefa
     * @throws com.stefanini.todolist.infrastructure.exceptions.ResourceNotFoundException
     *         se a tarefa não for encontrada
     */
    void deletar(Long id);
}

